crates/aquascope/src/analysis/boundaries/mod.rs

```diff
@@ -96,6 +114,129 @@ fn select_candidate_location<'tcx>(
   }
 }
 
+/// Return the constraints that occur nested within a [`HirId`].
+///
+/// Note, constraints involving regions belonging to the same SCC are removed.
+fn flow_constraints_at_hir_id<'a, 'tcx: 'a>(
+  ctxt: &'a PermissionsCtxt<'a, 'tcx>,
+  ir_mapper: &'a IRMapper<'a, 'tcx>,
+  hir_id: HirId,
+) -> Option<Vec<(Origin, Origin, Point)>> {
+  let mir_locations =
+    ir_mapper.get_mir_locations(hir_id, GatherDepth::Nested)?;
+
+  let all_constraints = mir_locations
+    .values()
+    .flat_map(|loc| {
+      let ps = ctxt.location_to_points(loc);
+      ctxt
+        .polonius_input_facts
+        .subset_base
+        .iter()
+        .filter(move |&(f, t, p)| {
+          !ctxt.is_universal_subset((*f, *t)) && ps.contains(p)
+        })
+        .copied()
+    })
+    .collect::<Vec<_>>();
+
+  Some(all_constraints)
+}
+
+/// If flow permissions are enabled, find expected flow permissions (if any) for the
+/// given `hir_id` within the larger `flow_context`.
+fn get_flow_permission<'tcx>(
+  ctxt: &PermissionsCtxt<'_, 'tcx>,
+  ir_mapper: &IRMapper<'_, 'tcx>,
+  flow_context: HirId,
+  hir_id: HirId,
+  span_to_range: impl Fn(Span) -> Range,
```

Comment:
Do we need to thread this closure everywhere? Can we construct an `AquascopeAnalysis` in the unit tests so we don't need two separate values for `span_to_range`?